db/view/view.cc

```diff
@@ -2360,6 +2475,205 @@ future<> view_builder::do_build_step() {
+        reader_concurrency_semaphore* semaphore;
+    };
+
+    distributed<replica::database>& _db;
+    table_id _table_id;
+    std::vector<reader_context> _contexts;
+public:
+    view_build_status_policy(distributed<replica::database>& db, table_id table_id)
+        : _db(db)
+        , _table_id(table_id)
+        , _contexts(smp::count) {
+    }
+    virtual mutation_reader create_reader(
+            schema_ptr schema,
+            reader_permit permit,
+            const dht::partition_range& range,
+            const query::partition_slice& slice,
+            tracing::trace_state_ptr trace_state,
+            mutation_reader::forwarding fwd_mr) override {
+        const auto shard = this_shard_id();
+        auto& cf = _db.local().find_column_family(_table_id);
+
+        _contexts[shard].range = make_foreign(make_lw_shared<const dht::partition_range>(range));
+        _contexts[shard].read_operation = make_foreign(std::make_unique<utils::phased_barrier::operation>(cf.read_in_progress()));
+        _contexts[shard].semaphore = &_db.local().get_reader_concurrency_semaphore();
+
+        return cf.as_mutation_source().make_reader_v2(
+                cf.schema(),
+                std::move(permit),
+                range,
+                slice,
+                std::move(trace_state),
+                streamed_mutation::forwarding::no,
+                fwd_mr);
+    }
+    virtual const dht::partition_range* get_read_range() const override {
+        const auto shard = this_shard_id();
+        return _contexts[shard].range.get();
+    }
+    virtual void update_read_range(lw_shared_ptr<const dht::partition_range> range) override {
+        const auto shard = this_shard_id();
+        _contexts[shard].range = make_foreign(std::move(range));
+    }
+    virtual future<> destroy_reader(stopped_reader reader) noexcept override {
+        auto ctx = std::move(_contexts[this_shard_id()]);
+        auto reader_opt = ctx.semaphore->unregister_inactive_read(std::move(reader.handle));
+        if  (!reader_opt) {
+            return make_ready_future<>();
+        }
+        return reader_opt->close().finally([ctx = std::move(ctx)] {});
+    }
+    virtual reader_concurrency_semaphore& semaphore() override {
+        const auto shard = this_shard_id();
+        if (!_contexts[shard].semaphore) {
+            auto& cf = _db.local().find_column_family(_table_id);
+            _contexts[shard].semaphore = &_db.local().get_reader_concurrency_semaphore();
+        }
+        return *_contexts[shard].semaphore;
+    }
+    virtual future<reader_permit> obtain_reader_permit(schema_ptr schema, const char* const description, db::timeout_clock::time_point timeout, tracing::trace_state_ptr trace_ptr) override {
+        auto& cf = _db.local().find_column_family(_table_id);
+        return semaphore().obtain_permit(schema, description, cf.estimate_read_memory_cost(), timeout, std::move(trace_ptr));
```

Comment:
it seems there is still an issue I need to look at.
when I do `SELECT COUNT(*) FROM system_distributed.view_build_status` it fails with timeout. I think it reaches the concurrency limit.

Reply:
I think there was a deadlock because we already have a reader permit and we try to obtain another one with the same semaphore
I changed it to a tracking only permit